implement grid-compose

eyad-hussein opened the pull request:
### Description

Building the grid-compose cli tool by implementing several commands adapted by docker-compose(up, down, version, etc)

### Changes

- [x] ps
- [x] up
- [x] down
- [x] version

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-go/issues/1022

### Checklist

- [ ] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstring

Omarabdul3ziz committed:
init grid-compose

Omarabdul3ziz committed:
add network support

Omarabdul3ziz committed:
use yaml.v3 to fix the custom unmarshal

eyad-hussein committed:
feat: implement init logic for ps command

xmonader commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,94 @@
+	rootCmd.AddCommand(downCmd)
+}
+
+type App struct {
```
Not the right place, should be in internal/app.go or similar

xmonader commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,94 @@
+	Specs  types.Specs
+}
+
+func NewApp(net, mne, filePath string) (App, error) {
```
factory function should be in internal/app.go and use better names eg mnemonic instead of mne

xmonader commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,94 @@
+	}, nil
+}
+
+func LoadSpecsFromFile(filePath string) (types.Specs, error) {
```
create a loadSpecsFromReader with io.Reader and don't bother about paths, opening/closing the files.

xmonader replied:
and of course, not the right place, internal/app.go

xmonader commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,94 @@
+	return specs, nil
+}
+
+func Execute() {
```
move this to the beginning of the file

xmonader commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,94 @@
+
+var (
+	app        App
+	configFile string
```
configPath

xmonader commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,94 @@
+	Short: "Grid-Compose is a tool for running multi-vm applications on TFGrid defined using a Yaml formatted file.",
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		var err error
+		app, err = NewApp(network, mnemonic, configFile)
```
if app, err := ... ; err != nil {
   log.Fatal ...
   }

xmonader commented on grid-compose/cmd/up.go:
```diff
@@ -0,0 +1,95 @@
+	},
+}
+
+func up(ctx context.Context) error {
```
move that into internal/app.go and make it a method

xmonader replied:
and same for the rest..

xmonader commented on grid-compose/cmd/up.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	for key, val := range results {
+		fmt.Printf("%s vm addresses:\n", key)
```
why mixing logging with printing?

xmonader commented on grid-compose/internal/utils.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/threefoldtech/zos/pkg/gridtypes/zos"
+)
+
+func GetProjectName(key string, twinId uint32) string {
```
method on the app

xmonader commented on grid-compose/internal/utils.go:
```diff
@@ -0,0 +1,41 @@
+	return fmt.Sprintf("compose/%v/%v", twinId, key)
+}
+
+func GetVmAddresses(vm workloads.VM) string {
```
move into vmaddresses.go in internal

xmonader commented on grid-compose/internal/utils.go:
```diff
@@ -0,0 +1,41 @@
+	return res
+}
+
+func GetRandomMyceliumIPSeed() ([]byte, error) {
```
move into internal/mycelium_seed.go

xmonader commented on grid-compose/pkg/types/types.go:
```diff
@@ -0,0 +1,70 @@
+type Service struct {
+	Flist       string       `yaml:"flist"`
+	Entrypoint  string       `yaml:"entrypoint,omitempty"`
+	Environment KVMap        `yaml:"environment"`
```
no need for the new type here map[string]string, specially if you don't have any new methods reacting on that type and doesn't provide any type safety

xmonader commented on grid-compose/pkg/types/types.go:
```diff
@@ -0,0 +1,70 @@
+
+type Service struct {
+	Flist       string       `yaml:"flist"`
+	Entrypoint  string       `yaml:"entrypoint,omitempty"`
```
this is required

xmonader commented on grid-compose/pkg/types/types.go:
```diff
@@ -0,0 +1,70 @@
+	Test     []string `yaml:"test"`
+	Interval string   `yaml:"interval"`
+	Timeout  string   `yaml:"timeout"`
+	Retries  int      `yaml:"retries"`
```
think of the type used, is that retries going to be a negative at anypoint? use uint or uint8 instead

xmonader commented on grid-compose/pkg/types/types.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+type HealthCheck struct {
+	Test     []string `yaml:"test"`
```
Is that a test command or what exactly?add a comment next to it

xmonader commented on grid-compose/pkg/types/types.go:
```diff
@@ -0,0 +1,70 @@
+
+type KVMap map[string]string
+
+func (m *KVMap) UnmarshalYAML(value *yaml.Node) error {
```
remove this

xmonader commented on grid-compose/cmd/down.go:
```diff
@@ -0,0 +1,28 @@
+		projectName := internal.GetProjectName(key, app.Client.TwinID)
+		log.Info().Str("projectName", projectName).Msg("canceling deployments")
+		if err := app.Client.CancelByProjectName(projectName); err != nil {
+			return err
```
wrap the error here was e.g `fmt.Errorf("failed to cancel project %s: %w", projectName, err)`

xmonader replied:
move to to a method on App struct in app.go

xmonader commented on grid-compose/cmd/down.go:
```diff
@@ -0,0 +1,28 @@
+	},
+}
+
+func down() error {
```
move to a method on App struct in internal/app.go

xmonader commented on grid-compose/cmd/ps.go:
```diff
@@ -0,0 +1,123 @@
+	},
+}
+
+func ps(ctx context.Context) error {
```
move to a method on the app

xmonader commented on grid-compose/cmd/ps.go:
```diff
@@ -0,0 +1,123 @@
+	// try to get contracts with the old project name format "<name>"
+	contracts, err := t.ContractsGetter.ListContractsOfProjectName(projectName, true)
+	if err != nil {
+		return workloads.Deployment{}, err
```
consider returning a wrapper error with more context

xmonader commented on grid-compose/cmd/version.go:
```diff
@@ -0,0 +1,17 @@
+	"github.com/spf13/cobra"
+)
+
+var version = "v0.0.1"
```
check how to properly do the versioning https://github.com/threefoldtech/tfgrid-sdk-go/blob/0c4a82b2e3fe68c8b70d4fac89bb3e31fecf8566/grid-cli/cmd/version.go

xmonader commented on grid-compose/pkg/types/types.go:
```diff
@@ -0,0 +1,70 @@
+package types
+
```
All of these fields require validation e.g valid flist url, positive disk space, valid node id, depends_on a service and so on.

xmonader commented on grid-compose/internal/utils.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+func GetVmAddresses(vm workloads.VM) string {
+	var res string
```
use a string Builder for this https://pkg.go.dev/strings#Builder

xmonader commented:
Please review the project structure, cobra commands should ONLY invoke methods defined on the app, don't leak any code logic into your commands. make sure the input is validated with decent error messages as well.

eyad-hussein committed:
enhance: implement logic to parse env vars and disks for up command

eyad-hussein committed:
refactor: refactor ps command implementation

xmonader commented:
still needs to handle the dependency resolution "and guard against cycles"

eyad-hussein committed:
refactor: refactor up and ps commands

eyad-hussein committed:
feat: implement deployment of two or more services on same network

eyad-hussein committed:
exmple: add more example files to cover test cases

eyad-hussein committed:
refactor: refactor logic for all commands to comply with cases and constraints

eyad-hussein committed:
docs: edit readme file

eyad-hussein committed:
docs: add supported cases for deployment, update readme

Omarabdul3ziz commented on grid-compose/cmd/down.go:
```diff
@@ -0,0 +1,20 @@
+	},
+}
+
+func init() {
```
move registering commands to the init() in root.go

Omarabdul3ziz commented on grid-compose/cmd/ps.go:
```diff
@@ -0,0 +1,24 @@
+
+func init() {
+	psCmd.PersistentFlags().BoolP("verbose", "v", false, "all information about deployed services")
+	psCmd.PersistentFlags().StringP("output", "o", "", "output result to a file")
```
we can drop this and simply use bash redirecting. `grid-compose ps > dep.json`

Omarabdul3ziz commented on grid-compose/cmd/ps.go:
```diff
@@ -0,0 +1,24 @@
+	Run: func(cmd *cobra.Command, args []string) {
+		flags := cmd.Flags()
+
+		if err := app.Ps(cmd.Context(), flags); err != nil {
```
invoke the flags here, and pass to the method only the values it needs

Omarabdul3ziz commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,44 @@
+
+var (
+	app        *internal.App
+	configPath string
```
to avoid the global variable, we could add the app to the context. it will shared between the commands

Omarabdul3ziz commented on grid-compose/cmd/root.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+func init() {
+	network = os.Getenv("NETWORK")
```
move getting the env to the Run in rootCmd

Omarabdul3ziz commented on grid-compose/internal/config/config.go:
```diff
@@ -0,0 +1,130 @@
+package config
```
this pkg could be named `parser` and could be in `pkg/` and should only be responsible for
- reading the config file
- one-to-one mapping to the config struct
- validate

another layer should do the mapping from the config struct to the deploymentData map with all the logic to create the missing fields 
- networks
- filter to get node id
- minimum resources if missing

Omarabdul3ziz commented on grid-compose/internal/config/config.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+// Config represents the configuration file content
+type Config struct {
```
define it with the other types

Omarabdul3ziz commented on grid-compose/internal/types/types.go:
```diff
@@ -0,0 +1,65 @@
+	Flist       string       `yaml:"flist"`
+	Entrypoint  string       `yaml:"entrypoint,omitempty"`
+	Environment []string     `yaml:"environment"`
+	Resources   Resources    `yaml:"resources"`
```
you can define a new type for this which implemnt the `YamlUnmarshlar` to directly have the environment as `map[string]string` which what zmachine.envirnment expect

Omarabdul3ziz commented on grid-compose/internal/utils/mycelium_seed.go:
```diff
@@ -0,0 +1,13 @@
+package utils
```
it is not a good practice to have a package named `utils`

instead add these methods in the package where it are used, okay to have a separate file for it in the same package.

Omarabdul3ziz commented on grid-compose/internal/utils/validation.go:
```diff
@@ -0,0 +1,20 @@
+
+import "github.com/cosmos/go-bip39"
+
+func ValidateCredentials(mnemonics, network string) bool {
```
better to be part of the parser/config layer/package

Omarabdul3ziz commented on grid-compose/internal/app.go:
```diff
@@ -0,0 +1,372 @@
+package internal
```
a better structure for this file could be a separate package `internal/app` have the `app.go` file for the app struct & factory method. and a file for each command with its utils `ps.go`, `down.go`, ...

Omarabdul3ziz commented on grid-compose/internal/app.go:
```diff
@@ -0,0 +1,372 @@
+
+// App is the main application struct that holds the client and the config data
+type App struct {
+	Client *deployer.TFPluginClient
```
is it needed to be public?

Omarabdul3ziz commented on grid-compose/internal/app.go:
```diff
@@ -0,0 +1,372 @@
+// App is the main application struct that holds the client and the config data
+type App struct {
+	Client *deployer.TFPluginClient
+	Config *config.Config
```
i think it is not used anywhere? if so remove it

Omarabdul3ziz commented on grid-compose/internal/app.go:
```diff
@@ -0,0 +1,372 @@
+
+				if !verbose {
+					if !dlAdded {
+						output.WriteString(fmt.Sprintf("%-15s | %-15s | %-15s | %-15s | %-10s | %s\n", contractDlData.Name, vm.NetworkName, vm.Name, vm.Mounts[0].DiskName, wl.Result.State, utils.GetVmAddresses(vm)))
```
could this formatting be abstracted? get the result as a map/array then pass it to some method maybe in `pkg/logging/print_resutl.go` it will be easier to make changes or even use a different pkg for formatting

Omarabdul3ziz commented on grid-compose/internal/app.go:
```diff
@@ -0,0 +1,372 @@
+		}
+
+		// freeCRU is not in NodeFilter?
+		var freeMRU, freeSRU uint64
```
yes freeCRU is not a filter. due to cores overprovisioning mulitple vms can use the same cpu cores, use totalCRU instead just making sure the node has the cores you need

Omarabdul3ziz reviewed (changes requested):
good job ya eyad 

i have some comments on the project structure

eyad-hussein committed:
feat: add support for depends_on optoin(dependency resolution)

refactor: refactor folder structure and add more examples

eyad-hussein committed:
feat: add support for health check option

refactor: refcator project structure

eyad-hussein committed:
docs: add future work file

eyad-hussein committed:
refactor: refactor project structure and code

tests: test each example manually

docs: populate readme file, add future_work, add config.md, and extend cases

eyad-hussein committed:
fix: add check for app type assertion, refactor: remove deploy package and move its components to app

Omarabdul3ziz committed:
Merge pull request #1155 from threefoldtech/development_compose_init_dependency_resolution

implementation of a docker-compose-like cli for managing deployment of services simultaneously

eyad-hussein committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-go into development_compose_init

eyad-hussein committed:
feat: add support for different storage units for service resources

refactor: move config to pkg and create storage parser

validation: add extensive validation for services and config

eyad-hussein committed:
fix: fix storage parser

eyad-hussein committed:
fix: change project name so type becomes Micro Virtual Machine in dashboard

xmonader commented:
Let's unblock this
